File: ui/trace-ui/src/types/StarfishEvents.ts
// Quantity Item
export interface QuantityItem {
    epc: string; // e.g., "urn:epc:class:lgtin:9506000.1233.a"
    quantity: number; // e.g., 12.0
    unit_of_measurement: string; // e.g., "kg"
}

// Base Event
export interface BaseEvent {
    event_time: string; // ISO timestamp, e.g., "2025-10-03T12:08:00.000Z"
    event_timezone_offset: string; // e.g., "+01:00"
}

// Creating Event
export interface CreatingEvent extends BaseEvent {
    eventType: "creating";
    biz_location: string; // e.g., "9506001112229"
    quantity_list: QuantityItem[];
}

// Shipping Event
export interface ShippingEvent extends BaseEvent {
    eventType: "shipping";
    ship_from: string;
    ship_to: string;
    items: QuantityItem[];
}

// Receiving Event
export interface ReceivingEvent extends BaseEvent {
    eventType: "receiving";
    shipped_from: string;
    received_at: string;
    items: QuantityItem[];
}

// Transforming Event
export interface TransformingEvent extends BaseEvent {
    eventType: "transforming";
    facility: string;
    transformation_id?: string | null;
    input_items: QuantityItem[];
    output_items: QuantityItem[];
}

// Packing Event
export interface PackingEvent extends BaseEvent {
    eventType: "packing";
    facility: string;
    container_id: string;
    input_items: QuantityItem[];
}

// Unpacking Event
export interface UnpackingEvent extends BaseEvent {
    eventType: "unpacking";
    facility: string;
    container_id: string;
    output_items: QuantityItem[];
}

// Discriminated union
export type StarfishEvent =
    | CreatingEvent
    | ShippingEvent
    | ReceivingEvent
    | TransformingEvent
    | PackingEvent
    | UnpackingEvent;

File: api/app/models/starfish_events.py
from pydantic import BaseModel, Field
from typing import List, Literal, Optional, Union

class QuantityItem(BaseModel):
    epc: str = Field(..., example="urn:epc:class:lgtin:9506000.1233.a")
    quantity: float = Field(..., example=12.0)
    unit_of_measurement: str = Field(..., example="kg")

class BaseEvent(BaseModel):
    event_time: str = Field(..., example="2025-10-03T12:08:00.000Z")
    event_timezone_offset: str = Field(..., example="+01:00")

# Creating event
class CreatingEvent(BaseEvent):
    eventType: Literal["creating"] = "creating"
    biz_location: str = Field(..., example="9506001112229")
    quantity_list: List[QuantityItem] = Field(..., min_length=1)

# Shipping event
class ShippingEvent(BaseEvent):
    eventType: Literal["shipping"] = "shipping"
    ship_from: str
    ship_to: str
    items: List[QuantityItem] = Field(..., min_length=1)

# Receiving event
class ReceivingEvent(BaseEvent):
    eventType: Literal["receiving"] = "receiving"
    shipped_from: str
    received_at: str
    items: List[QuantityItem] = Field(..., min_length=1)

# Transforming event
class TransformingEvent(BaseEvent):
    eventType: Literal["transforming"] = "transforming"
    facility: str
    transformation_id: Optional[str] = None
    input_items: List[QuantityItem] = Field(..., min_length=1)
    output_items: List[QuantityItem] = Field(..., min_length=1)

# Packing event
class PackingEvent(BaseEvent):
    eventType: Literal["packing"] = "packing"
    facility: str
    container_id: str
    input_items: List[QuantityItem] = Field(..., min_length=1)

# Unpacking event
class UnpackingEvent(BaseEvent):
    eventType: Literal["unpacking"] = "unpacking"
    facility: str
    container_id: str
    output_items: List[QuantityItem] = Field(..., min_length=1)

StarfishEvent = Union[
    CreatingEvent,
    ShippingEvent,
    ReceivingEvent,
    TransformingEvent,
    PackingEvent,
    UnpackingEvent,
]

